lumeer-core/src/main/java/io/lumeer/core/facade/ViewFacade.java

```diff
@@ -0,0 +1,173 @@
+import javax.inject.Inject;
+
+@RequestScoped
+public class ViewFacade {
+
+   @Inject
+   private AuthenticatedUser authenticatedUser;
+
+   @Inject
+   private PermissionsChecker permissionsChecker;
+
+   @Inject
+   private ViewDao viewDao;
+
+   @Inject
+   private WorkspaceKeeper workspaceKeeper;
+
+   public ViewFacade() {
+   }
+
+   ViewFacade(AuthenticatedUser authenticatedUser, PermissionsChecker permissionsChecker, ViewDao viewDao, WorkspaceKeeper workspaceKeeper) {
+      this.authenticatedUser = authenticatedUser;
+      this.permissionsChecker = permissionsChecker;
+      this.viewDao = viewDao;
+      this.workspaceKeeper = workspaceKeeper;
+   }
+
+   @PostConstruct
+   public void initViewDao() {
+      if (workspaceKeeper.getProject().isPresent()) {
+         viewDao.setProject(workspaceKeeper.getProject().get());
+      }
+   }
+
+   public View createView(View view) {
+      // TODO check collection permissions
+
+      // TODO generate view code if not provided
+
+      Permission defaultUserPermission = new SimplePermission(authenticatedUser.getUserEmail(), View.ROLES);
+      view.getPermissions().updateUserPermissions(defaultUserPermission);
+
+      return viewDao.createView(view);
+   }
+
+   public View updateView(final String code, final View view) {
+      View storedView = viewDao.getViewByCode(code);
+      permissionsChecker.checkRole(storedView, Role.MANAGE);
+
+      keepStoredPermissions(view, storedView.getPermissions());
+      View updatedView = viewDao.updateView(storedView.getId(), view);
+
+      return keepOnlyActualUserRoles(updatedView);
+   }
+
+   private void keepStoredPermissions(View view, Permissions storedPermissions) {
+      Set<Permission> userPermissions = storedPermissions.getUserPermissions();
+      view.getPermissions().updateUserPermissions(userPermissions.toArray(new Permission[0]));
+
+      Set<Permission> groupPermissions = storedPermissions.getGroupPermissions();
+      view.getPermissions().updateGroupPermissions(groupPermissions.toArray(new Permission[0]));
+   }
+
+   public void deleteView(final String code) {
+      View view = viewDao.getViewByCode(code);
+      permissionsChecker.checkRole(view, Role.MANAGE);
+
+      viewDao.deleteView(view.getId());
+   }
+
+   public View getViewByCode(final String code) {
+      View view = viewDao.getViewByCode(code);
+      permissionsChecker.checkRole(view, Role.READ);
+
+      return keepOnlyActualUserRoles(view);
+   }
+
+   public List<View> getAllViews() {
+      return viewDao.getAllViews().stream()
+                    .filter(view -> permissionsChecker.hasRole(view, Role.READ))
+                    .map(this::keepOnlyActualUserRoles)
+                    .collect(Collectors.toList());
+   }
+
+   private View keepOnlyActualUserRoles(View view) {
+      Set<Role> roles = permissionsChecker.getActualRoles(view);
+      Permission permission = new SimplePermission(authenticatedUser.getUserEmail(), roles);
```

Comment:
`getActualRoles` method returns all roles assigned to the currently logged in user regardless of whether they are direct or gained through the group membership. `ViewFacade` then returns a resource with only one permission record (for the authenticated user) where there are all these roles so they can be easily checked on the front-end.